File: tests/org/elixir_lang/reference/callable/Issue354Test.java
package org.elixir_lang.reference.callable;

import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiPolyVariantReference;
import com.intellij.psi.PsiReference;
import com.intellij.psi.ResolveResult;
import com.intellij.psi.impl.source.tree.LeafPsiElement;
import org.elixir_lang.PlatformTestCase;
import org.elixir_lang.psi.ElixirIdentifier;
import org.elixir_lang.psi.call.Call;

public class Issue354Test extends PlatformTestCase {
    /*
     * Tests
     */

    public void testLoggerLogstashBackend() {
        myFixture.configureByFile("logger_logstash_backend.ex");
        PsiElement elementAtCaret = myFixture.getFile().findElementAt(myFixture.getCaretOffset());

        assertNotNull(elementAtCaret);
        assertInstanceOf(elementAtCaret, LeafPsiElement.class);

        PsiElement parent = elementAtCaret.getParent();

        assertNotNull(parent);
        assertInstanceOf(parent, ElixirIdentifier.class);

        PsiElement grandParent = parent.getParent();

        assertNotNull(grandParent);
        assertInstanceOf(grandParent, Call.class);

        Call grandParentCall = (Call) grandParent;
        PsiReference reference = grandParentCall.getReference();

        assertNotNull(reference);

        assertInstanceOf(reference, PsiPolyVariantReference.class);

        PsiPolyVariantReference psiPolyVariantReference = (PsiPolyVariantReference) reference;

        ResolveResult[] resolveResults = psiPolyVariantReference.multiResolve(true);

        assertEquals(resolveResults.length, 2);

        ResolveResult firstResolveResult = resolveResults[0];

        assertTrue(firstResolveResult.isValidResult());

        PsiElement firstElement = firstResolveResult.getElement();

        assertNotNull(firstElement);
        assertEquals("%{line: line, port: port} = context", firstElement.getParent().getText());

        ResolveResult secondResolveResult = resolveResults[1];

        assertTrue(secondResolveResult.isValidResult());

        PsiElement secondElement = secondResolveResult.getElement();

        assertNotNull(secondElement);
        assertEquals("context = %{backend: true}", secondElement.getParent().getText());
    }

    /*
     * Protected Instance Methods
     */

    @Override
    protected String getTestDataPath() {
        return "testData/org/elixir_lang/reference/callable/issue_354";
    }
}

File: tests/org/elixir_lang/reference/callable/Issue354Test.kt
package org.elixir_lang.reference.callable

import org.elixir_lang.psi.ElixirIdentifier
import com.intellij.psi.PsiPolyVariantReference
import com.intellij.psi.impl.source.tree.LeafPsiElement
import org.elixir_lang.PlatformTestCase
import org.elixir_lang.psi.call.Call

/**
 * `context` on the right of `%{line: line, port: port} = context` reads the parameter, so a later `context` resolves
 * to the parameter alone and not to that read as a rebinding.
 */
class Issue354Test : PlatformTestCase() {
    fun testLoggerLogstashBackend() {
        myFixture.configureByFile("logger_logstash_backend.ex")

        val elementAtCaret = myFixture.file.findElementAt(myFixture.caretOffset)
        assertNotNull(elementAtCaret)
        assertInstanceOf(elementAtCaret, LeafPsiElement::class.java)

        val parent = elementAtCaret!!.parent
        assertNotNull(parent)
        assertInstanceOf(parent, ElixirIdentifier::class.java)

        val grandParent = parent.parent
        assertNotNull(grandParent)
        assertInstanceOf(grandParent, Call::class.java)
        val grandParentCall = grandParent as Call

        val reference = grandParentCall.reference
        assertNotNull(reference)
        assertInstanceOf(reference, PsiPolyVariantReference::class.java)
        val psiPolyVariantReference = reference as PsiPolyVariantReference

        val resolveResults = psiPolyVariantReference.multiResolve(true)
        assertEquals(1, resolveResults.size)

        val resolveResult = resolveResults[0]
        assertTrue(resolveResult.isValidResult)
        assertEquals("context = %{backend: true}", resolveResult.element!!.parent.text)
    }

    /*
     * Protected Instance Methods
     */
    override fun getTestDataPath(): String = "testData/org/elixir_lang/reference/callable/issue_354"
}
